src/Interpreter.js

```diff
@@ -86,15 +86,57 @@ export default class Interpreter {
                 // We're at the end, nothing to do
                 resolve();
             } else {
-                this.doCommand(programSequence.getCurrentProgramStep()).then(() => {
-                    // When the command has completed, increment
-                    // the programCounter and resolve the step Promise
-                    this.app.incrementProgramCounter(() => {
-                        resolve();
+                const currentProgramStep = programSequence.getCurrentProgramStep();
+                const command = currentProgramStep.block;
+                if (command === 'startLoop') {
+                    const { iterationsLeft, label } = currentProgramStep;
+                    if (iterationsLeft != null && label != null && iterationsLeft > 0) {
+                        this.app.decrementLoopIterations(label, () => {
+                            resolve();
+                        });
+                    }
+                } else if (command === 'endLoop') {
+                    const { label } = currentProgramStep;
+                    if( label != null ) {
+                        const currentProgram = programSequence.getProgram();
+                        let programCounter = programSequence.getProgramCounter();
+                        let isNested = false;
+                        for (let i = 0; i < programSequence.getProgramCounter(); i++) {
```

Comment:
An alternative algorithm:

* Start iterating through the program backwards from the current program counter
* Until we reach the corresponding startLoop
* Then stop
* Then we don't need the `isNested` variable: if we encounter a startLoop that isn't the corresponding one, we reset its `iterationsLeft`, otherwise, we have reached as far as we need to go